tf_quant_finance/math/jacobian_test.py

```diff
@@ -0,0 +1,120 @@
+
+@test_util.run_all_in_graph_and_eager_modes
+class JacobianTest(parameterized.TestCase, tf.test.TestCase):
+  """Jacobian test cases."""
+
+  @parameterized.named_parameters(
+    {
+      'testcase_name': 'SinglePrecision',
+      'dtype': tf.float32
+    }, {
+      'testcase_name': 'DoublePrecision',
+      'dtype': tf.float64
+    },
+  )
+  def test_jacobian(self, dtype):
+    """Test function jacobian."""
+    # Shape [2]
+    x = tf.range(1, 3, dtype=dtype)
+    func = partial(quadratic, x=x)
+    with self.subTest("SingleTensor"):
+      # Shape [3]
+      ps = tf.constant([1.0, 2.0, -1.0], dtype=dtype)
+      # Shape [2, 3]
+      expected_jacobian = [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]]
+      jacobian = self.evaluate(tff.math.jacobian(func, ps))
+      self.assertEqual(jacobian.shape, (2, 3))
+      np.testing.assert_allclose(jacobian, expected_jacobian)
+    with self.subTest("BatchedTensor"):
+      # Shape [2, 3]
+      ps = tf.constant([[1.0, 2.0, -1.0], [2.0, 0.0, 0.0]], dtype=dtype)
+      # Shape [2, 3]
+      expected_jacobian = [
+        [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]],
+        [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]],
+      ]
+      jacobian = self.evaluate(tff.math.jacobian(func, ps))
+      self.assertEqual(jacobian.shape, (2, 2, 3))
+      np.testing.assert_allclose(jacobian, expected_jacobian)
+
+  @parameterized.named_parameters(
+    {
+      'testcase_name': 'SinglePrecision',
+      'dtype': tf.float32
+    }, {
+      'testcase_name': 'DoublePrecision',
+      'dtype': tf.float64
+    },
+  )
+  def test_value_and_jacobian(self, dtype):
+    """Test function value_and_jacobian."""
+    # Shape [2]
+    x = tf.range(1, 3, dtype=dtype)
+    with self.subTest("SingleTensor"):
+      func = partial(quadratic, x=x)
+      # Shape [3]
+      ps = tf.constant([1.0, 2.0, -1.0], dtype=dtype)
+      # Shape [2]
+      expected_values = [2.0, 7.0]
+      # Shape [2, 3]
+      expected_jacobian = [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]]
+      values, jacobian = self.evaluate(
+          tff.math.value_and_jacobian(func, ps))
+      self.assertEqual(values.shape, (2,))
+      self.assertEqual(jacobian.shape, (2, 3,))
+      np.testing.assert_allclose(values, expected_values)
+      np.testing.assert_allclose(jacobian, expected_jacobian)
+    with self.subTest("BatchedTensor"):
+      func = partial(quadratic, x=x)
+      # Shape [2, 3]
+      ps = tf.constant([[1.0, 2.0, -1.0], [2.0, 0.0, 0.0]], dtype=dtype)
+      # Shape [2, 2]
+      expected_values = [[2.0, 7.0], [2.0, 8.0]]
+      # Shape [2, 2, 3]
+      expected_jacobian = [
+        [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]],
+        [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]],
+      ]
+      values, jacobian = self.evaluate(
+          tff.math.value_and_jacobian(func, ps))
+      self.assertEqual(values.shape, (2, 2,))
+      self.assertEqual(jacobian.shape, (2, 2, 3))
+      np.testing.assert_allclose(values, expected_values)
+      np.testing.assert_allclose(jacobian, expected_jacobian)
```

Comment:
It is easier to spot a bug if subtest is used. See, e.g., 
https://github.com/google/tf-quant-finance/blob/96fb93058020f4dd0f2aedf0d1c8d34e77fad887/tf_quant_finance/models/geometric_brownian_motion/geometric_brownian_motion_test.py#L47

Reply:
Addressed in 8a357ce